pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -431,40 +434,55 @@ async def run(self):
         """Build `DeclarativeContent` from the repodata."""
         progress_data = dict(message="Downloading Metadata Files", code="sync.downloading.metadata")
         with ProgressReport(**progress_data) as metadata_pb:
-            self.data.metadata_pb = metadata_pb
-
             downloader = self.remote.get_downloader(
-                url=urlpath_sanitize(self.data.remote_url, "repodata/repomd.xml")
+                url=urlpath_sanitize(self.remote_url, "repodata/repomd.xml")
             )
             result = await downloader.run()
             metadata_pb.increment()
 
             repomd_path = result.path
-            self.data.repomd = cr.Repomd(repomd_path)
+            repomd = cr.Repomd(repomd_path)
 
-            self.repository.last_sync_revision_number = self.data.repomd.revision
+            self.repository.last_sync_revision_number = repomd.revision
             self.repository.last_sync_repomd_checksum = get_sha256(repomd_path)
 
-            await self.parse_distribution_tree()
-            await self.parse_repository_metadata()
-            await self.parse_modules_metadata()
-            await self.parse_packages_components()
-            await self.parse_content()
+            checksum_types = {}
+            metadata_downloaders = {}
+            types_to_download = (
+                set(PACKAGE_REPODATA)
+                | set(UPDATE_REPODATA)
+                | set(COMPS_REPODATA)
+                | set(MODULAR_REPODATA)
+            )
+
+            # TODO: do we actually need to download all metadata for mirror case, or use RepoMetadataFile?
+            for record in repomd.records:
+                checksum_types[record.type] = getattr(CHECKSUM_TYPES, record.checksum_type.upper())
+                record.checksum_type = getattr(CHECKSUM_TYPES, record.checksum_type.upper())
 
-            # now send modules down the pipeline since all relations have been set up
-            for modulemd in self.data.modulemd_list:
-                await self.put(modulemd)
+                if record.type in types_to_download:
+                    sanitized = urlpath_sanitize(self.remote_url, record.location_href)
+                    downloader = self.remote.get_downloader(url=sanitized)
+                    metadata_downloaders[record.type] = downloader.run()
 
-            for dc_group in self.data.dc_groups:
-                await self.put(dc_group)
+            self.repository.original_checksum_types = checksum_types
+
+            # TODO: do the results get updated as shallow copies, or do we need to re-associate them?
+            # TODO: increment metadata progress report counter
+            try:
+                await asyncio.gather(*list(metadata_downloaders.values()))
```

Comment:
Side benefit: we actually get to download the metadata in parallel properly now.